core/docker/build.sh

```diff
@@ -71,13 +76,23 @@ TAG_PREFIX="trino:${TRINO_VERSION}"
 
 for arch in "${ARCHITECTURES[@]}"; do
     echo "🫙  Building the image for $arch"
-    docker build \
-        "${WORK_DIR}" \
-        --pull \
-        --platform "linux/$arch" \
-        -f Dockerfile \
-        -t "${TAG_PREFIX}-$arch" \
+    BUILD_OPTIONS=(
+        "${WORK_DIR}"
+        --platform "linux/$arch"
+        -f Dockerfile
+        -t "${TAG_PREFIX}-$arch"
         --build-arg "TRINO_VERSION=${TRINO_VERSION}"
+        )
+    if [ -n "${BUILD_CACHE_IMAGE}" ]; then
+        echo "🎯 Using ${BUILD_CACHE_IMAGE} as Docker build cache"
+        docker pull \
+            --platform "linux/$arch" \
+            "${BUILD_CACHE_IMAGE}"
+        BUILD_OPTIONS+=(--cache-from "${BUILD_CACHE_IMAGE}")
```

Comment:
I think this isn't working because the final image is missing intermediate layers. Instead of using the published image, we could start exporting the cache to a file using `--cache-to` and saving it in GitHub Actions cache.

Reply:
Whoa, `gha` is supported out of the box! https://docs.docker.com/engine/reference/commandline/buildx_build/#cache-to
```bash
docker buildx build --cache-to=type=gha .
```

Reply:
> I think this isn't working because the final image is missing intermediate layers. Instead of using the published image, we could start exporting the cache to a file using `--cache-to` and saving it in GitHub Actions cache.
> 
> Uploading and downloading the cache will make the workflow run longer, but should be slightly compensated by faster image builds and this is the price of increasing reliability. If we're concerned with this cache size, we can try excluding the Trino server from it.
> 
> This solution should still be better than publishing a base image since we won't have to care when to rebuild it.



> I think this isn't working because the final image is missing intermediate layers. Instead of using the published image, we could start exporting the cache to a file using `--cache-to` and saving it in GitHub Actions cache.
> 
> Uploading and downloading the cache will make the workflow run longer, but should be slightly compensated by faster image builds and this is the price of increasing reliability. If we're concerned with this cache size, we can try excluding the Trino server from it.
> 
> This solution should still be better than publishing a base image since we won't have to care when to rebuild it.

Removing intermediate layers by default is one of the features of buildx, but it can be disabled with `mode=max`.

Reply:
> Whoa, `gha` is supported out of the box! https://docs.docker.com/engine/reference/commandline/buildx_build/#cache-to
> 
> ```shell
> docker buildx build --cache-to=type=gha .
> ```

I tried it, but I see no cache entries being generated.

Reply:
Docker docs link to this page: https://github.com/tonistiigi/go-actions-cache/blob/master/api.md#authentication
Looks like we should make sure that `ACTIONS_RUNTIME_TOKEN` env var is set. I don't see any warnings in the logs, though.

I found other docs here for using `docker/build-push-action`: https://docs.docker.com/build/ci/github-actions/examples/#github-cache

Reply:
It worked!
![image](https://user-images.githubusercontent.com/795177/215481766-ea4e6d2b-de57-4d51-8856-7ffd7722072f.png)

Can you push an empty commit, so we can see if the cache will be correctly used?

Reply:
> It worked! ![image](https://user-images.githubusercontent.com/795177/215481766-ea4e6d2b-de57-4d51-8856-7ffd7722072f.png)
> 
> Can you push an empty commit, so we can see if the cache will be correctly used?

I'm concerned this will make out 10GB cache unusable since a single PR pipeline can now:

* upload 1.5GB of Maven dependencies
* upload a 1GB of Docker image layers for each supported platform

Reply:
It should be fine as long as the entries for master are not evicted too often.